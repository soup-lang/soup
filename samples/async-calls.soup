# This simulates a regular async function
func asyncDiv(x, y, callback)
    setTimeout(():
        if y = 0
            return callback(Error('Divide ' + x + ' by zero'), undefined)
        else
            callback(null, x / y)
    , 0)

console.log(asyncDiv(5, 1, %%))
console.log(asyncDiv(5, 2, %%))
console.log(asyncDiv(5, 4, %%))

try
    asyncDiv(5, 0, %%)
catch e
    console.log(e.message)

# ===========================================
console.log('-- Non-regular async functions')

func asyncTextGen(input, cb)
    setTimeout(%, 0)
    if input.length < 4
        cb(input + ' is quite short')
    else
        cb(input + ' is quite long')

asyncTextGen('cat', %text)
console.log(text)
asyncTextGen(text, %moreText)
console.log(moreText)

# ====================================
console.log('-- Regular async lambda')
regularAsyncTextGen: (input, %%):
    setTimeout(%, 0)
    return input + (' is quite short' if input.length < 4 else ' is quite long')

console.log(regularAsyncTextGen('cat', %%))
console.log(regularAsyncTextGen('raven', %%))

# =================================================
console.log('-- Raise exception at async function')
func asyncRaiseError(%%)
    setTimeout(%, 0)
    throw Error('error message')

try
    asyncRaiseError(%%)
catch e
    console.log(e.message)

console.log('End')
