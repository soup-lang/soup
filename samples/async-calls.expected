5
2.5
1.25
Divide 5 by zero
-- Non-regular async functions
cat is quite short
cat is quite short is quite long
-- Regular async lambda
cat is quite short
raven is quite long
-- Raise exception at async function
error message
End
